### ISSUE TITLE
server reports "Unable to locate the project file" after upgrading ts-standard to 12.0.1

### ISSUE THREAD

**Issue by brenc** [chenxsan/vscode-standardjs#483]
**What version of this package are you using?**
12.0.1

**What operating system, Node.js, and npm version?**
Debian 11 x64, Node v16.15.1, npm v8.11.0.

**What happened?**
After upgrading from `ts-standard` 11 to 12.0.1, I now get this in VS Code:

[Info  - 11:00:40 PM] TypeScript Standard Style server stopped.
[Info  - 11:00:43 PM] TypeScript Standard Style server is running.
Unable to locate the project file. A project file (tsconfig.json or tsconfig.eslint.json) is required in order to use ts-standard.
[Info  - 11:00:43 PM] Connection to server got closed. Server will restart.
[Info  - 11:00:43 PM] TypeScript Standard Style server stopped.
[Info  - 11:00:43 PM] TypeScript Standard Style server stopped.

It tries to start the server several times then gives up. I can run `ts-standard` on the CLI and it works fine. `tsconfig.json` is there of course so I'm not sure what's going on. Downgrading back to 11 fixes this. 

**What did you expect to happen?**
For the server to start up.

**Are you willing to submit a pull request to fix this bug?**
Sure.

**Comment by theoludwig** [comment]
Hello, thanks for your report!
I see that many people have the same issue.

Unfortunately, I can't reproduce it, it works fine on my side. (ref: https://github.com/standard/ts-standard/issues/272#issuecomment-1287148476).
Please provide a minimal way to reproduce the issue, also take a look of the example given here: https://github.com/standard/vscode-standard/tree/master/client/testFixture/ts-standard to know how to configure `ts-standard`.

**Comment by hrmorley34** [comment]
I've also been experiencing this.

- Extension version: v2.1.2
- ts-standard: 12.0.1
- Windows 10, Node v16.13.2, npm 8.19.2

Steps to reproduce:
- I downloaded the test fixture linked (`package.json`, `tsconfig.json`, `src/lint-errors.ts`, and `src/no-lint-errors.ts`), and opened in Code
- Ran `npm i`
- Opened `lint-errors.ts`, and the server failed to start.

I then disabled every extension other than this one in the workspace, reloaded, and the error occured again.

<details>
<summary>Logs</summary>

```
[Info  - 14:41:10] JavaScript Standard Style server stopped.
[Info  - 14:41:15] JavaScript Standard Style server is running.
Unable to locate the project file. A project file (tsconfig.json or tsconfig.eslint.json) is required in order to use ts-standard.
[Info  - 14:41:16] Connection to server got closed. Server will restart.
[Info  - 14:41:16] JavaScript Standard Style server stopped.
[Info  - 14:41:16] JavaScript Standard Style server stopped.
[Info  - 14:41:18] JavaScript Standard Style server is running.
Unable to locate the project file. A project file (tsconfig.json or tsconfig.eslint.json) is required in order to use ts-standard.
[Info  - 14:41:19] Connection to server got closed. Server will restart.
[Info  - 14:41:19] JavaScript Standard Style server stopped.
[Info  - 14:41:19] JavaScript Standard Style server stopped.
[Info  - 14:41:21] JavaScript Standard Style server is running.
Unable to locate the project file. A project file (tsconfig.json or tsconfig.eslint.json) is required in order to use ts-standard.
[Info  - 14:41:22] Connection to server got closed. Server will restart.
[Info  - 14:41:22] JavaScript Standard Style server stopped.
[Info  - 14:41:22] JavaScript Standard Style server stopped.
[Info  - 14:41:24] JavaScript Standard Style server is running.
Unable to locate the project file. A project file (tsconfig.json or tsconfig.eslint.json) is required in order to use ts-standard.
[Info  - 14:41:24] Connection to server got closed. Server will restart.
[Info  - 14:41:24] JavaScript Standard Style server stopped.
[Info  - 14:41:24] JavaScript Standard Style server stopped.
[Info  - 14:41:26] JavaScript Standard Style server is running.
Unable to locate the project file. A project file (tsconfig.json or tsconfig.eslint.json) is required in order to use ts-standard.
[Error - 14:41:27] The JavaScript Standard Style server crashed 5 times in the last 3 minutes. The server will not be restarted. See the output for more information.
[Info  - 14:41:27] JavaScript Standard Style server stopped.
```

</details>

Running `npx ts-standard`/`npm run lint` in the terminal still works fine.

```
$ npx ts-standard
npm WARN config init.author.name Use `--init-author-name` instead.
ts-standard: Standard for TypeScript! (https://github.com/standard/ts-standard)
ts-standard: Run `ts-standard --fix` to automatically fix some problems.
  C:\Users\henry\AppData\Local\Temp\tmp.kk6i5dX7G7\src\lint-errors.ts:1:13: Strings must use singlequote. (@typescript-eslint/quotes)
  C:\Users\henry\AppData\Local\Temp\tmp.kk6i5dX7G7\src\lint-errors.ts:1:29: Extra semicolon. (@typescript-eslint/semi)
```

As above, after running `npm i -D ts-standard@11.0.0` and reloading Code, it works as expected.

![image](https://user-images.githubusercontent.com/10192165/197543668-09c04eb7-d921-4922-859e-2dbace6932d6.png)

<details>
<summary>Logs</summary>

```
[Info  - 14:57:57] JavaScript Standard Style server stopped.
[Info  - 14:58:03] JavaScript Standard Style server is running.
[Info  - 14:58:05] TypeScript Standard Style library loaded from: c:\Users\henry\AppData\Local\Temp\tmp.kk6i5dX7G7\node_modules\ts-standard\dist\index.js
TypeError: callback is not a function
    at c:\Users\henry\.vscode\extensions\standard.vscode-standard-2.1.2\server\out\server.js:667:17
    at process.processTicksAndRejections (node:internal/process/task_queues:96:5)
TypeError: callback is not a function
    at c:\Users\henry\.vscode\extensions\standard.vscode-standard-2.1.2\server\out\server.js:667:17
    at process.processTicksAndRejections (node:internal/process/task_queues:96:5)
TypeError: callback is not a function
    at c:\Users\henry\.vscode\extensions\standard.vscode-standard-2.1.2\server\out\server.js:667:17
    at process.processTicksAndRejections (node:internal/process/task_queues:96:5)
TypeError: callback is not a function
    at c:\Users\henry\.vscode\extensions\standard.vscode-standard-2.1.2\server\out\server.js:667:17
    at process.processTicksAndRejections (node:internal/process/task_queues:96:5)
```

</details>

**Comment by vileppanen** [comment]
Reproducing for me also.

OS Windows 11
Extension version v2.1.2
ts-standard version 12.0.1
VSCode version 1.72.2

My project is bootstrapped with CRA using TS template --> root folder contains `tsconfig.json` & the `package.json` with the following included:
```
  "ts-standard": {
    "project": "./tsconfig.json",
    "env": [
      "node"
    ]
  }
``` 

When I open the folder in VSCode and open up any `.tsx` file in the editor --> the output console starts throwing similar errors indicating the issues finding `tsconfig.json` & eventually leading the standard server crashing:

```
[Info  - 20.19.01] JavaScript Standard Style server stopped.
[Info  - 20.19.01] JavaScript Standard Style server is running.
Unable to locate the project file. A project file (tsconfig.json or tsconfig.eslint.json) is required in order to use ts-standard.
[Info  - 20.19.02] Connection to server got closed. Server will restart.
[Info  - 20.19.02] JavaScript Standard Style server stopped.
[Info  - 20.19.02] JavaScript Standard Style server stopped.
[Info  - 20.19.02] JavaScript Standard Style server is running.
Unable to locate the project file. A project file (tsconfig.json or tsconfig.eslint.json) is required in order to use ts-standard.
[Info  - 20.19.03] Connection to server got closed. Server will restart.
[Info  - 20.19.03] JavaScript Standard Style server stopped.
[Info  - 20.19.03] JavaScript Standard Style server stopped.
[Info  - 20.19.03] JavaScript Standard Style server is running.
Unable to locate the project file. A project file (tsconfig.json or tsconfig.eslint.json) is required in order to use ts-standard.
```

**Comment by theoludwig** [comment]
Thanks for your reproductions.
It would already be good if someone could start by making a failing test in https://github.com/standard/vscode-standard/tree/master/client/testFixture/ts-standard.
The tests seem to pass without problems, and I still can't reproduce the problem on my side. :confused: 

Also if someone knows how to fix the problem, we would love a PR. :heart:

**Comment by vileppanen** [comment]
> Thanks for your reproductions. It would already be good if someone could start by making a failing test in https://github.com/standard/vscode-standard/tree/master/client/testFixture/ts-standard. The tests seem to pass without problems, and I still can't reproduce the problem on my side. 😕
> 
> Also if someone knows how to fix the problem, we would love a PR. ❤️

On Node v16.17.0 & NPM 8.15.0

1. I cloned the the vscode-standard repo.
2. Navigated to the `client/testFixture/ts-standard` folder
3. installed dependencies via `npm i`
4. Opened `src/lint-errors.ts` in the editor
5. Started experiencing the `Unable to locate the project file. A project file (tsconfig.json or tsconfig.eslint.json) is required in order to use ts-standard.`

**Comment by vileppanen** [comment]
I started looking at the latest changes in `ts-standard`, and was curious about this:
https://github.com/standard/ts-standard/commit/c3de7fe08c1803f31ead1cd94b1a09fdc78c2194#diff-e7ad5f13058c7d8fe78b1910deadef9c6f714595824f73d12f572e8b7e9d258d

If I run `npm run lint` in the `vscode-standard/client/testFixture/ts-standard` --> the linting works just fine from the command line. However, the VSCode editor side seems to be the issue. This led me to wonder, does the `process.cwd()` point to somewhere else than the project root, when executed in the context of VSCode?

**Comment by theoludwig** [comment]
@vileppanen I don't know what else to do. It's weird that so many people have this issue and still e2e tests are green, and I can't reproduce it, I really don't know what could be the cause. Maybe I'm missing a step or something from your reproduction, I don't really know.
![Screenshot from 2022-10-27 21-05-01](https://user-images.githubusercontent.com/25207499/198377236-41f412e0-5e9e-4aad-919e-a62f162beabf.png)

> This led me to wonder, does the process.cwd() point to somewhere else than the project root, when executed in the context of VSCode?

The `process.cwd()` in VSCode is the path of the opened folder (same behavior as the CLI really), so this should be fine. Compared to v11, there should not be so many differences for this part.

**Comment by pedro-w** [comment]
I have also seen this on 
```
Version: 1.73.0 (user setup)
Commit: 8fa188b2b301d36553cbc9ce1b0a146ccb93351f
Date: 2022-11-01T15:34:06.111Z
Electron: 19.0.17
Chromium: 102.0.5005.167
Node.js: 16.14.2
V8: 10.2.154.15-electron.0
OS: Windows_NT x64 10.0.22621
Sandboxed: No
```
I think [the comment](https://github.com/standard/vscode-standard/issues/483#issuecomment-1293913611) by @vileppanen might be the cause. I really struggled to get the VS code debugger to bind any breakpoints to the server code so I resorted to `console.log`... I can see that at this point:

https://github.com/standard/ts-standard/blob/01879c3c32cb06185c57fe922e0849c052b4be38/resolve-tsconfig.js#L57

the passed in `cwd` is actually `C:\Users\USERNAME\AppData\Local\Programs\Microsoft VS Code` and not the project dir. I haven't got a linux set-up but I'm guessing this is a difference between platforms.

Apart from that, there may be legitimate circumstances when there is no project file, and in this case it would be better for the server to print a warning and do nothing rather than crashing as it does now, because this triggers VS code's restart behaviour which is not pleasant for the user.

Hope that helps, let me know and I can do some more testing.

**Comment by pedro-w** [comment]
I have also encountered the same crash on MacOS (to add to the reports on Windows and Debian). I made a minimal repo which will show this.
https://github.com/pedro-w/ts-std-test
There are 'instructions' in the README there. Maybe I'm doing something wrong in my setup?

```
Version: 1.73.0
Commit: 8fa188b2b301d36553cbc9ce1b0a146ccb93351f
Date: 2022-11-01T15:38:50.881Z
Electron: 19.0.17
Chromium: 102.0.5005.167
Node.js: 16.14.2
V8: 10.2.154.15-electron.0
OS: Darwin x64 19.6.0
Sandboxed: No
```

**Comment by pedro-w** [comment]
This patch will stop the error for me but I do not know enough about the code to say whether it is a true fix or just a hack
```
diff --git a/server/src/server.ts b/server/src/server.ts
index 45b5457..7b152cf 100644
--- a/server/src/server.ts
+++ b/server/src/server.ts
@@ -427,8 +427,15 @@ async function resolveSettings (
         async path => {
           let library = path2Library.get(path)
           if (library == null) {
+            // Need to load the library dynamically, making sure CWD is temporarily
+            // set to the target workspace
+            const oldCwd = process.cwd()
+            if (settings.workspaceFolder != null) {
+              process.chdir(settings.workspaceFolder.uri.fsPath)
+            }
             // eslint-disable-next-line no-new-func, @typescript-eslint/no-implied-eval
             library = (await Function(`return import('file://${path.replace(/\\/g, '\\\\')}')`)()).default
+            process.chdir(oldCwd)
             if (library?.lintText == null) {
               settings.validate = false
               connection.console.error(
```

**Comment by theoludwig** [comment]
Thanks for your help and debugging! @pedro-w 
Could you make a PR with your patch and preferably with a failing e2e test, and then make it so that the test succeeds with your patch?

**Comment by theoludwig** [comment]
Thank you everyone for your help!
The issue should be resolved since [v2.1.3](https://github.com/standard/vscode-standard/releases/tag/v2.1.3) that just released.
Please reopen an issue if the issue persists. :smile:
